cpp/include/raft/neighbors/detail/ivf_flat_build.cuh

```diff
@@ -416,4 +420,152 @@ inline void fill_refinement_index(raft::resources const& handle,
+                                                            index.data_ptrs().data_handle(),
+                                                            index.inds_ptrs().data_handle(),
+                                                            index.dim(),
+                                                            index.veclen(),
+                                                            index.n_lists(),
+                                                            max_indice,
+                                                            ptrs_to_data.data());
+  RAFT_CUDA_TRY(cudaPeekAtLastError());
+
+  auto n_reconstruction = vector_ids.extent(0);
+  const dim3 block_dim2(256);
+  const dim3 grid_dim2(raft::ceildiv<size_t>(n_reconstruction, block_dim2.x));
+  reconstruct_batch_kernel<<<grid_dim2, block_dim2, 0, stream>>>(vector_ids.data_handle(),
+                                                                 (const T**)ptrs_to_data.data(),
+                                                                 index.dim(),
+                                                                 index.veclen(),
+                                                                 n_reconstruction,
+                                                                 vector_out.data_handle());
+  RAFT_CUDA_TRY(cudaPeekAtLastError());
+}
+
+template <typename T>
+__global__ void reconstruct_list_data_kernel(
+  T* out_vectors,
+  T* in_list_data,
+  std::variant<uint32_t, const uint32_t*> offset_or_indices,
+  uint32_t len,
+  size_t veclen,
+  uint32_t dim)
+{
+  for (uint32_t ix = threadIdx.x + blockDim.x * blockIdx.x; ix < len; ix += blockDim.x) {
+    const uint32_t src_ix = std::holds_alternative<uint32_t>(offset_or_indices)
+                              ? std::get<uint32_t>(offset_or_indices) + ix
+                              : std::get<const uint32_t*>(offset_or_indices)[ix];
+
+    using group_align         = Pow2<kIndexGroupSize>;
+    const uint32_t group_ix   = group_align::div(src_ix);
+    const uint32_t ingroup_ix = group_align::mod(src_ix) * veclen;
+
+    for (uint32_t l = 0; l < dim; l += veclen) {
+      for (uint32_t j = 0; j < veclen; j++) {
+        out_vectors[ix * dim + l + j] = in_list_data[l * kIndexGroupSize + ingroup_ix + j];
+      }
+    }
+  }
+}
+
+/** Decode the list data; see the public interface for the api and usage. */
+template <typename T, typename IdxT>
+void reconstruct_list_data(raft::resources const& handle,
+                           const index<T, IdxT>& index,
+                           device_matrix_view<T, IdxT, row_major> out_vectors,
+                           uint32_t label,
+                           uint32_t offset)
+{
+  auto stream = raft::resource::get_cuda_stream(handle);
+
+  uint32_t len = out_vectors.extent(0);
+  const dim3 block_dim(256);
+  const dim3 grid_dim(raft::ceildiv<size_t>(len, block_dim.x));
```

Comment:
A nitpick: use the newer [raft::div_rounding_up_safe](https://github.com/rapidsai/raft/blob/f34cdbd740d23dae06ac88938094b9429cb81e4e/cpp/include/raft/util/integer_utils.hpp#L116C23-L116C43) here and above. Perhaps, we should deprecate `raft::ceildiv` in favor of `raft::div_rounding_up_unsafe`?...